app/controllers/api/base_controller.rb

```diff
@@ -37,11 +37,18 @@ def verify_mfa_requirement
     render plain: "Gem requires MFA enabled; You do not have MFA enabled yet.", status: :forbidden
   end
 
+  def verify_api_key_gem_scope
+    return unless @api_key.rubygem && @api_key.rubygem != @rubygem
+
+    render plain: "You do not have permission. This API key is scoped to #{@api_key.rubygem.name}.", status: :forbidden
+  end
+
   def authenticate_with_api_key
     params_key = request.headers["Authorization"] || ""
     hashed_key = Digest::SHA256.hexdigest(params_key)
     @api_key   = ApiKey.find_by_hashed_key(hashed_key)
-    render_unauthorized unless @api_key
+    return render_unauthorized unless @api_key
+    render_invalid_api_key unless @api_key.valid?
```

Comment:
I am not sure if I understand in which case will the valid? check come into play. Please add a test for it if there is. If you have already added one, please make sure to update the wording to reflect that valid? is being checked.

Reply:
Yes, that would be the use case. I agree it's awkward to be using `valid?` in this case. I created a [method](https://github.com/rubygems/rubygems.org/blob/7eecd2094dd0700e094639bf317ef45d9427290a/app/models/api_key.rb#L38) under api keys for this authorization behaviour. Let me know if you had something else in mind.